### DISCUSSION TITLE
Lombok support

### DISCUSSION THREAD

**Discussion by DarrMirr** [eclipse-theia/theia#9070] [Q&A] [answered]
Hi there!

I have a question. Does theia-ide supports lombok?

My steps:
1. I pull theia-full docker image and run it. 
2. I open my Java project at theia-ide. 
3. I successfully compile my project using terminal command `mvn compile`
4. I face to compilation error while running application from theia-ide
5. I investigate compilation errors and find out that theia could not infer lombok annotations.

I have found issue [#2363](https://github.com/eclipse-theia/theia/issues/2363) but I could not understand how to add lombok library. I try to add this library at launch.json but it does not work.

So, does theia-ide supports lombok? Is there any docs how to plug lombok to theia-ide?

Best regards
Darr Mirr

**Comment by vince-fugnitto** [comment, marked as answer]
@DarrMirr thank you for the discussion, in order to support `lombok` in your custom application you can make use of the `vscode-lombok` plugin: 
- https://github.com/GabrielBB/vscode-lombok

You can either reference the plugin under `theiaPlugins` of your application's package.json and it will download it given the link, or you can put it manually under your plugin folder for test purposes. At this time the plugin is not available under the open-source [open-vsx](https://open-vsx.org/) registry, but you can request to add it if you believe it works well.

**Reply by DarrMirr** [reply]
@vince-fugnitto thank you for response!

I have faced with trouble to install `lombok` to theia ide at docker. 

I am Java developer and do not familiar with javascript environment and visial studio one.

I have read some articles like [how-to-add-extensions-and-plugins-to-eclipse-theia](https://eclipsesource.com/blogs/2019/10/17/how-to-add-extensions-and-plugins-to-eclipse-theia/) in order to install `lombok` using theia ide tools. But I do not succeed to do this due to docs is difficult to understand for me.

So I manage with plugin installation at own way. I leave this docs here. It maybe help for someone.

**How to install any plugin manually at theia docker**

1. Run docker instance as root `sudo docker run --user=root -d -p 3000:3000 --name theiaide -v "$(pwd):/home/project:cached" theiaide/theia-full`
1. Open webide at your browser (For example, http://localhost:3000)
1. Click `Terminal` menu item and then click `New terminal` submenu item
1. Click left button mouse to terminal window in order to make it active window
1. Update OS repository cache executing command `apt-get update`
1. Install 7z by executing command `apt-get install p7zip-full`
1. Create new plugin directory at theia default plugins one (For example, `mkdir -p /home/theia/plugins/lombok`)
1. Move to created directory (For example `cd /home/theia/plugins/lombok `)
1. Download `vsix` file. (For example, `curl "https://marketplace.visualstudio.com/_apis/public/gallery/publishers/GabrielBB/vsextensions/vscode-lombok/1.0.1/vspackage" > lombok.vsix`)
1. Uzip vsix file by 7z (For example, `7z x lombok.vsix`)
1. Uzip unziped file (For example, `unzip lombok`). 
1. Close tab with webide at your browser
1. Restart docker container `sudo docker restart theiaide`
1. Open webide at your browser (For example, http://localhost:3000)
1. Click `View` menu item and then click `Extensions` submenu item
1. Open `Built-in` extensions list
1. Find your plugin in list (Lombok is installed extension at my case)
![image](https://user-images.githubusercontent.com/17032652/108043647-63dd2a80-7052-11eb-938c-92d397333a3f.png)
